lib/pacer_test.go

```diff
@@ -0,0 +1,148 @@
+		// 1 hit/sec, 1 hit sent, 1s elapsed => 1s until next hit
+		{1, time.Second, time.Second, 1, time.Second, false},
+		// 1 hit/sec, 2 hits sent, 1s elapsed => 2s until next hit
+		{1, time.Second, time.Second, 2, 2 * time.Second, false},
+		// 1 hit/sec, 10 hits sent, 1s elapsed => 10s until next hit
+		{1, time.Second, time.Second, 10, 10 * time.Second, false},
+		// 1 hit/sec, 10 hits sent, 11s elapsed => 0s until next hit
+		{1, time.Second, 11 * time.Second, 10, 0, false},
+		// 2 hit/sec, 9 hits sent, 4.9s elapsed => 100ms until next hit
+		{2, time.Second, (49 * time.Second) / 10, 9, 100 * time.Millisecond, false},
+	} {
+		r := &Rate{Freq: tt.freq, Per: tt.per}
+
+		wait, stop := r.Pace(tt.elapsed, tt.count)
+		if have, want := wait, tt.wait; have != want {
+			t.Errorf("test case %d: %+v: wait=%v, want %v", i, r, have, want)
+		}
+		if have, want := stop, tt.stop; have != want {
+			t.Errorf("test case %d: %+v: stop=%v, want %v", i, r, have, want)
+		}
+	}
+}
+
+var quarterPeriods = map[string]float64{
+	"MeanUp":   MeanUp,
+	"Peak":     Peak,
+	"MeanDown": MeanDown,
+	"Trough":   Trough,
+}
+
+type sineTest struct {
+	p, m, a int
+}
+
+func (st sineTest) Pacer(startAt float64) SinePacer {
+	return SinePacer{
+		Period:  time.Duration(st.p) * time.Second,
+		Mean:    Rate{st.m, time.Second},
+		Amp:     Rate{st.a, time.Second},
+		StartAt: startAt,
+	}
+}
+
+func (st sineTest) AmpHits() float64 {
+	return float64(st.a*st.p) / (2 * math.Pi)
+}
+
+func (st sineTest) Hits(frac, startAt float64) uint64 {
+	return uint64(math.Round(
+		float64(st.m*st.p)*frac +
+			st.AmpHits()*(math.Cos(startAt)-math.Cos(startAt+frac*2*math.Pi))))
+}
+
+func (st sineTest) Nanos(frac, startAt float64) time.Duration {
+	return time.Duration(1 / (float64(st.m) + float64(st.a)*math.Sin(startAt+frac*2*math.Pi)))
+}
+
+func TestSinePacerHits(t *testing.T) {
+	tests := []sineTest{
+		{20 * 60, 100, 90},
+		{60, 1000, 10},
+		{1, 10, 7},
+		{1, 1, 0},
+		// These test cases failed with off-by-one errors before applying
+		// math.Round in Hits, due to floating-point maths differences.
+		{1e6, 10, 7},
+		{60, 1000, 999},
+	}
+
+	for i, test := range tests {
+		for name, sa := range quarterPeriods {
+			sr := test.Pacer(sa)
+			if got, want := sr.Hits(sr.Period/4), test.Hits(0.25, sa); got != want {
+				t.Errorf("%d(%s): hits after 1/4 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(sr.Period/2), test.Hits(0.5, sa); got != want {
+				t.Errorf("%d(%s): hits after 1/2 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(3*sr.Period/4), test.Hits(0.75, sa); got != want {
+				t.Errorf("%d(%s): hits after 3/4 period = %d, want %d", i, name, got, want)
+			}
+			if got, want := sr.Hits(sr.Period), test.Hits(1, sa); got != want {
+				t.Errorf("%d(%s): hits after full period = %d, want %d", i, name, got, want)
+			}
+		}
+	}
+}
+
+func TestSinePacerFlat(t *testing.T) {
```

Comment:
Don't forget edge cases also

Reply:
These tests are pretty hacky, I wrote them quickly but relied on "manual integration testing" (i.e. push it and see if it does the right thing) to validate things mostly. I am a terrible person ;-)

I agree that for the final PR they need a lot of work.

Reply:
@LasTshaMAN: It seems that you are quite inspired on the testing front. Would you be up to contribute this part of the work?

Reply:
Yeah, I like testing things :)

Sure, I'll try my best, just let me know which form of contribution (as @fluffle mentioned below) is preferable.

Reply:
Ok, saw the comment below, a follow-up PR would be convenient for me to do. 

Let's finish up with this one and then I'll get to work on testing new `Pacer` functionality (as long as I have spare time of course, this May I'm a little busy ...).